# Урок 2. Основы Python: переменные, типы и условия

**Имя и фамилия: Опёнов Данила**  
**Группа: 25-ИИбо-1**

Это единый Notebook для практики на занятии и домашней работы. Работайте в
нём во время урока, сохраните результат, а после выполнения домашней части
загрузите тот же файл как `homework/lesson-02/solution.ipynb`.

## 1. Проверяем рабочий цикл

Выполните ячейку. Под числом сформулируйте словами, что оно означает.

In [41]:
1250 * 3

3750

**Мой вывод:**

## 2. Переменные и состояние Kernel

`=` присваивает значение имени. Для нескольких слов используем `snake_case`.
Python определяет тип значения во время выполнения.

In [42]:
product_name = "Наушники"
price = 1250
quantity = 3
order_total = price * quantity

print(product_name, order_total)

Наушники 3750


### Практика 1. Изменяем заказ

1. До запуска спрогнозируйте результат.
2. Создайте свои `price` и `quantity`.
3. Рассчитайте `order_total`.
4. Измените количество и выполните ячейку повторно.
5. Исправьте имя `order_totall`, не запуская неверный вариант повторно.

In [43]:
# Напишите решение здесь.
order_totall = "Зонтики"
price = 1999
quantity = 14
order_total = price * quantity
print(order_totall, order_total)

Зонтики 27986


**Что изменилось и почему:**

## 3. Типы и преобразования

`int` — целое число, `float` — дробное, `str` — текст, `bool` — логическое
значение, `None` — значение пока отсутствует. `type()` проверяет тип.

In [44]:
product_name = "Клавиатура"
price = 3200
discount_rate = 0.05
is_member = True
promo_code = None

type(product_name), type(price), type(discount_rate), type(is_member), type(promo_code)

(str, int, float, bool, NoneType)

In [45]:
quantity_text = "2"
quantity = int(quantity_text)
order_total = price * quantity

print(order_total, type(quantity_text), type(quantity))

6400 <class 'str'> <class 'int'>


### Практика 2. Исправляем тип

Код `"Цена: " + 3200` вызывает ошибку. Исправьте его двумя способами:
через `str()` и через f-строку. Объясните причину ошибки.

In [46]:
# price = 3200
price = 3200
result_1 = "Цена: " + str(price)
print(result_1)

Цена: 3200


In [47]:
price = 3200
result_2 = f"Цена: {price}"
print(result_2)

Цена: 3200


**Причина ошибки: Ошибка возникает из-за того, что в Python нельзя напрямую складывать (конкатенировать) строку (str) и число (int) через оператор +.**

## 4. Строки и f-строки

f-строка вставляет значения и результаты выражений внутрь понятного текста.

In [48]:
product_name = "Клавиатура"
price = 3200
quantity = 2
order_total = price * quantity

print(f"{quantity} × {product_name}: {order_total} рублей")

2 × Клавиатура: 6400 рублей


### Практика 3. Понятное сообщение

Выведите одной f-строкой название, цену, количество, сумму и наличие
клубной карты. Измените данные и обновите результат.

In [49]:
# Напишите решение здесь.
name = "Мышка"
cost = 1999
order = 5
total = cost * order
carta = False
print(f"Товар:{name}, Цена:{cost}, Кол-во:{order}, Стоимость:{total}, Наличие карты:{carta}")

Товар:Мышка, Цена:1999, Кол-во:5, Стоимость:9995, Наличие карты:False


## 5. Сравнения и логические операции

`= ` присваивает, `==` сравнивает. Операторы `!=`, `>`, `<`, `>=`, `<=`
возвращают `True` или `False`. Условия объединяются через `and`, `or`, `not`.

In [50]:
free_delivery_threshold = 5000

print(4999 >= free_delivery_threshold)
print(5000 >= free_delivery_threshold)
print(5001 >= free_delivery_threshold)

False
True
True


### Практика 4. Предсказываем результат

Проверьте правило: скидка доступна, если сумма не меньше 5000 **и** есть
клубная карта. Используйте три заданных набора:

1. `order_total = 4999`, `is_member = True`;
2. `order_total = 5000`, `is_member = False`;
3. `order_total = 5000`, `is_member = True`.

Перед запуском запишите прогноз `True` или `False`, затем проверьте его.

Исправьте ошибку `if order_total = 5000:` и объясните замену оператора.

In [51]:
# Выполните проверки здесь.
order_total = 4999
is_member = True
has_discount = order_total >= 5000 and is_member
print(f"Набор 1: {has_discount}")
order_total = 5000
is_member = False
has_discount = order_total >= 5000 and is_member
print(f"Набор 2: {has_discount}")
order_total = 5000
is_member = True
has_discount = order_total >= 5000 and is_member
print(f"Набор 3: {has_discount}")

Набор 1: False
Набор 2: False
Набор 3: True


**Мой вывод о границе и составном условии: Ошибка в выражении if order_total = 5000 заключается в использовании оператора присваивания = вместо оператора сравнения == или проверки условия «не меньше» >=.**

## 6. `if` / `elif` / `else`

Python проверяет ветви сверху вниз и выполняет первую подходящую. После
условия ставится двоеточие, команды внутри блока имеют одинаковый отступ.

In [52]:
order_total = 6500

if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 5000:
    discount_rate = 0.05
else:
    discount_rate = 0

total_after_discount = order_total * (1 - discount_rate)
print(f"Скидка: {discount_rate:.0%}; после скидки: {total_after_discount:.2f} руб.")

Скидка: 5%; после скидки: 6175.00 руб.


### Практика 5. Исправляем ошибки

Перенесите фрагмент в рабочую ячейку и исправьте двоеточие, отступы и
недостающую ветвь. Затем проверьте 4999, 5000, 6999 и 7000.

```python
if order_total >= 7000
discount_rate = 0.10
elif order_total >= 5000:
    discount_rate = 0.05
```

In [53]:
# Исправленный код и проверки.
def check_discount(order_total):
  if order_total >= 7000:
    discount_rate = 0.10
  elif order_total >= 5000:
    discount_rate = 0.05
  else:
    discount_rate = 0.00
  print(f"Сумма: {order_total} -> Скидка: {discount_rate * 100}%")
check_discount(4999)
check_discount(5000)
check_discount(6999)
check_discount(7000)

Сумма: 4999 -> Скидка: 0.0%
Сумма: 5000 -> Скидка: 5.0%
Сумма: 6999 -> Скидка: 5.0%
Сумма: 7000 -> Скидка: 10.0%


**Какие ошибки были исправлены: пропщуенный знак, отступ, логическая ошибка**

## 7. Самостоятельная практика «Итог заказа»

**Исходные правила:** скидка 10% при сумме от 7000; скидка 5% при сумме
от 5000 и наличии клубной карты; иначе скидка 0%. Доставка бесплатна, если
сумма после скидки не меньше 5000, иначе доставка стоит 300 рублей.

**Что сделать:**

1. Создайте `product_name`, `price`, `quantity`, `is_member`.
2. Рассчитайте `order_total`.
3. Через `if` / `elif` / `else` присвойте `discount_rate`.
4. Рассчитайте `total_after_discount`.
5. Через второе условие присвойте `actual_delivery_cost`.
6. Рассчитайте `final_total` и выведите все результаты одной f-строкой.
7. Проверьте три сценария: без скидки; со скидкой 5%; ровно 7000 до скидки.
8. Для каждого сценария заполните подготовленное поле ниже.

**Готово, если:** код выполняется без ошибок, значения меняются внутри
ветвей, проверены три сценария и вывод соответствует результату.

In [54]:
# 1. Исходные данные
product_name = "Ковёр"
price = 700
quantity = 10
is_member = False
# 2. Исходная сумма
order_total = price * quantity

# 3. Выбор discount_rate
if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 5000 and is_member:
    discount_rate = 0.05
else:
    discount_rate = 0.0

# 4. Сумма после скидки
total_after_discount = order_total * (1 - discount_rate)

# 5. Выбор actual_delivery_cost
if total_after_discount >= 5000:
    actual_delivery_cost = 0
else:
    actual_delivery_cost = 300

# 6. Итоговая сумма и f-строка
final_total = total_after_discount + actual_delivery_cost

output_string = (
    f"Товар: {product_name} | Сумма: {order_total} руб. | "
    f"Скидка: {discount_rate * 100:.0f}% | После скидки: {total_after_discount} руб. | "
    f"Доставка: {actual_delivery_cost} руб. | Итого к оплате: {final_total} руб."
)
print(output_string)

Товар: Ковёр | Сумма: 7000 руб. | Скидка: 10% | После скидки: 6300.0 руб. | Доставка: 0 руб. | Итого к оплате: 6300.0 руб.


### Результаты проверок

**Сценарий 1: price = 400 quantity = 10 is_member = False**  
**Результат и вывод: Товар: Ковёр | Сумма: 4000 руб. | Скидка: 0% | После скидки: 4000.0 руб. | Доставка: 300 руб. | Итого к оплате: 4300.0 руб.**

**Сценарий 2: price = 400 quantity = 14 is_member = True**  
**Результат и вывод: Товар: Ковёр | Сумма: 5600 руб. | Скидка: 5% | После скидки: 5320.0 руб. | Доставка: 0 руб. | Итого к оплате: 5320.0 руб.**

**Сценарий 3 — граничное значение: price = 700 quantity = 10
is_member = False**  
**Результат и вывод: Товар: Ковёр | Сумма: 7000 руб. | Скидка: 10% | После скидки: 6300.0 руб. | Доставка: 0 руб. | Итого к оплате: 6300.0 руб.**

## 8. Кратко: `input()` и `while`

`input()` всегда возвращает строку. Для числа требуется преобразование:

```python
quantity = int(input("Количество: "))
```

Этот пример не запускаем через `Run All`, чтобы Notebook не ожидал ввод.
`while` повторяет блок, пока условие истинно; условие должно изменяться.

In [55]:
attempt = 0
max_attempts = 3

while attempt < max_attempts:
    print(f"Попытка {attempt + 1}")
    attempt += 1

Попытка 1
Попытка 2
Попытка 3


### Найдите риск бесконечного цикла

Почему код ниже не завершится? Что нужно изменить?

```python
attempt = 0
while attempt < 3:
    print(attempt)
```

**Мой ответ: attempt не увеличивается из-за этого будет выполняться бесконечно т.к никогда не достигнет финального значения**

# Домашняя работа

Продолжайте работать в этом же Notebook. После выполнения загрузите его как
`homework/lesson-02/solution.ipynb`.

## Мини-исследование «Какие правила заказа выгодны покупателю?»

Все задания выполняются **ниже в уже подготовленных ячейках**. Новые ячейки
для обязательной части создавать не требуется.

**Обязательный результат:** базовый расчёт и четыре заданных эксперимента.
Для каждого эксперимента сохраните код, числовой результат и вывод словами.
Затем выполните минимум два задания расширенного уровня.

## Вопрос и гипотеза

**Мой вопрос:  При каких условиях изменения правил (повышение порога скидки или бесплатной доставки) покупатель несёт наибольшие или наименьшие затраты?**  
**Моя гипотеза: Увеличение скидки наиболее выгодно покупателю, а увеличение порога бесплатной доставки сильнее всего ударит по заказам средней стоимости.**

In [56]:
# Задание 1. Базовая модель заказа
product_name = "Наушники"
price = 1400
quantity = 3
is_member = True
discount_threshold = 5000
discount_rate = 0.10
free_delivery_threshold = 6000
delivery_cost = 400

# Рассчитайте order_total.
order_total = price * quantity
# Через if / else присвойте applied_discount_rate.
if order_total >= discount_threshold:
    applied_discount_rate = discount_rate
else:
    applied_discount_rate = 0.0
# Рассчитайте total_after_discount.
total_after_discount = order_total * (1 - applied_discount_rate)
# Через второе if / else присвойте actual_delivery_cost.
if total_after_discount >= free_delivery_threshold:
    actual_delivery_cost = 0
else:
    actual_delivery_cost = delivery_cost
#Рассчитайте final_total и выведите понятную f-строку.
final_total = total_after_discount + actual_delivery_cost

print(
    f"Заказ: {product_name} x {quantity} шт. | Исходная сумма: {order_total} руб. | "
    f"Применённая скидка: {applied_discount_rate * 100:.0f}% | "
    f"Доставка: {actual_delivery_cost} руб. | Итого к оплате: {final_total} руб."
)

Заказ: Наушники x 3 шт. | Исходная сумма: 4200 руб. | Применённая скидка: 0% | Доставка: 400 руб. | Итого к оплате: 4600.0 руб.


## Обязательные эксперименты

В каждом блоке используйте правила базовой модели. Меняйте только указанные
данные, выполняйте расчёт полностью и заполняйте вывод под Code-ячейкой.

In [57]:
# Эксперимент 1: ниже порога скидки
# Используйте price = 1200, quantity = 4.
# Рассчитайте скидку, доставку и final_total.
price = 1200
quantity = 4
order_total = price * quantity
if order_total >= 5000:
    applied_discount_rate = 0.10
else:
    applied_discount_rate = 0.0

total_after_discount = order_total * (1 - applied_discount_rate)
if total_after_discount >= 6000:
    actual_delivery_cost = 0
else:
    actual_delivery_cost = 400

final_total = total_after_discount + actual_delivery_cost

print(
    f"Сумма: {order_total} | Скидка: {applied_discount_rate*100}% | "
    f"После скидки: {total_after_discount} | Доставка: {actual_delivery_cost} | Итого: {final_total}"
)

Сумма: 4800 | Скидка: 0.0% | После скидки: 4800.0 | Доставка: 400 | Итого: 5200.0


**Вывод 1:** Какая скидка и доставка применились? Почему? Какова итоговая сумма?  Применилась скидка 0%, так как сумма заказа (4800 руб.) оказалась ниже установленного порога в 5000 руб. Доставка составила 400 руб., поскольку сумма после скидки (4800 руб.) не достигла порога бесплатной доставки в 6000 руб. Какова итоговая сумма? Итоговая сумма равна 5200 руб.

In [58]:
# Эксперимент 2: ровно на пороге скидки
# Используйте price = 1000, quantity = 5.
# Проверьте, применяется ли скидка при order_total == discount_threshold.
price = 1000
quantity = 5
order_total = price * quantity
if order_total >= 5000:
    applied_discount_rate = 0.10
else:
    applied_discount_rate = 0.0

total_after_discount = order_total * (1 - applied_discount_rate)
print(
    f"Сумма: {order_total} | Примененная скидка: {applied_discount_rate*100}% | После скидки: {total_after_discount}"
)

Сумма: 5000 | Примененная скидка: 10.0% | После скидки: 4500.0


**Вывод 2:** Какой оператор включил границу? Что изменится при замене >= на >? Границу включил оператор >= (больше или равно). Скидка применилась, так как условие 5000 >= 5000 истинно (True). При замене оператора с >= на > пограничное значение 5000 перестанет включаться в условие скидки: скидка станет 0%, а сумма после скидки останется равной 5000 руб.

In [59]:
# Эксперимент 3: граница бесплатной доставки
# Последовательно проверьте total_after_discount: 5999, 6000 и 6001.
# Для каждого значения определите actual_delivery_cost и final_total.
test_values = [5999, 6000, 6001]
delivery_cost = 400
for total_after_discount in test_values:
    if total_after_discount >= 6000:
        actual_delivery_cost = 0
    else:
        actual_delivery_cost = 400

    final_total = total_after_discount + actual_delivery_cost
    print(
        f"После скидки: {total_after_discount} руб. -> Доставка: {actual_delivery_cost} руб. | Итого: {final_total} руб."
    )

После скидки: 5999 руб. -> Доставка: 400 руб. | Итого: 6399 руб.
После скидки: 6000 руб. -> Доставка: 0 руб. | Итого: 6000 руб.
После скидки: 6001 руб. -> Доставка: 0 руб. | Итого: 6001 руб.


**Вывод 3:** Для каких значений доставка бесплатна? Что происходит точно на границе? Доставка бесплатна для значений 6000 и 6001. Точно на границе (6000) происходит включение условия бесплатной доставки, благодаря чему стоимость доставки падает с 400 руб. до 0 руб., а итоговая сумма к оплате становится равна ровно 6000 руб. (в то время как при 5999 руб. покупатель заплатит больше — 6399 руб.).

## Эксперимент 4. Конфликт правил доставки

Используйте `order_total = 6500`, скидку 10% и порог бесплатной доставки
6000 рублей. Сравните два правила:

1. право на бесплатную доставку определяется **до скидки**;
2. право на бесплатную доставку определяется **после скидки**.

Рассчитайте две итоговые суммы.

In [60]:
order_total = 6500
discount_rate = 0.10
free_delivery_threshold = 6000
delivery_cost = 400

# Рассчитайте total_after_discount.
total_after_discount = order_total * (1 - discount_rate)
# Правило 1: определите delivery_before_discount и итоговую сумму.
if order_total >= free_delivery_threshold:
    delivery_before_discount = 0
else:
    delivery_before_discount = delivery_cost
final_total_rule1 = total_after_discount + delivery_before_discount
# Правило 2: определите delivery_after_discount и итоговую сумму.
if total_after_discount >= free_delivery_threshold:
    delivery_after_discount = 0
else:
    delivery_after_discount = delivery_cost
final_total_rule2 = total_after_discount + delivery_after_discount

print(f"Правило 1 (до скидки): Итого к оплате = {final_total_rule1} руб. (Доставка: {delivery_before_discount} руб.)")
print(f"Правило 2 (после скидки): Итого к оплате = {final_total_rule2} руб. (Доставка: {delivery_after_discount} руб.)")

Правило 1 (до скидки): Итого к оплате = 5850.0 руб. (Доставка: 0 руб.)
Правило 2 (после скидки): Итого к оплате = 6250.0 руб. (Доставка: 400 руб.)


**Вывод 4:** Какое правило выгоднее покупателю? Какое — магазину? Какое правило вы считаете более понятным и почему? Правило 1 выгоднее покупателю, так как итоговая сумма меньше (5850 руб. против 6250 руб.). Правило 2 выгоднее магазину, так как оно вынуждает клиента оплачивать доставку, если скидка опустила чек ниже порога. Более понятным я считаю Правило 1, потому что покупатель сразу видит в корзине, набрал ли он товаров на бесплатную доставку, и его не ожидает неприятный сюрприз в виде внезапно ставшей платной доставки после применения скидки.

## Общий вывод

Ответьте не менее чем тремя предложениями: что вы проверили, какие границы
оказались важны и какое правило сильнее всего повлияло на итоговую сумму.

**Мой общий вывод: В ходе мини-исследования мы проверили, как пороговые значения скидок и стоимости доставки влияют на конечную сумму чека. Мы выяснили, что строго на границах (например, ровно 5000 или 6000 рублей) оператор >= играет решающую роль, включая эти суммы в льготные условия. Сильнее всего на итоговую сумму повлияло правило расчёта доставки после скидки: из-за него покупатель может потерять всю выгоду от скидки и заплатить больше, чем планировал.**

## Дополнительный уровень

Выполните **минимум два** задания. Для каждого уже подготовлены отдельные
Code- и Markdown-ячейки ниже.

1. Три уровня скидки и четыре граничных проверки.
2. Клубная карта: сравнение `and` и `or`.
3. Промокод и правило совместного применения скидок.
4. Собственное правило магазина и минимум три проверки.

In [66]:
# Расширение 1. Три уровня скидки
# До 3000 — 0%; от 3000 до 6999 — 5%; от 7000 — 10%.
# Проверьте 2999, 3000, 6999 и 7000 без циклов и списков.
order_total = 2999
if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0.0
print(f"Сумма: {order_total} -> Скидка: {discount_rate*100}%")

# Проверка значения 3000
order_total = 3000
if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0.0
print(f"Сумма: {order_total} -> Скидка: {discount_rate*100}%")

# Проверка значения 6999
order_total = 6999
if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0.0
print(f"Сумма: {order_total} -> Скидка: {discount_rate*100}%")

# Проверка значения 7000
order_total = 7000
if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0.0
print(f"Сумма: {order_total} -> Скидка: {discount_rate*100}%")

Сумма: 2999 -> Скидка: 0.0%
Сумма: 3000 -> Скидка: 5.0%
Сумма: 6999 -> Скидка: 5.0%
Сумма: 7000 -> Скидка: 10.0%


**Вывод расширения 1: Проверка показала точное переключение скидок на граничных значениях. На суммах 2999 и 6999 скидки составили 0% и 5% соответственно. В точках 3000 и 7000 скидки успешно переключились на более высокий уровень (5% и 10%), подтверждая корректность граничных условий >=.**

In [67]:
# Расширение 2. Клубная карта
order_total = 4500
is_member = True

# Сравните два правила: достаточная сумма AND карта; достаточная сумма OR карта.
threshold = 5000
rule_and = (order_total >= threshold) and is_member
rule_or = (order_total >= threshold) or is_member

print(f"Результат правила AND: {rule_and}")
print(f"Результат правила OR: {rule_or}")


Результат правила AND: False
Результат правила OR: True


**Вывод расширения 2:** Когда правила дают разные результаты и почему? Правила дают разные результаты, когда выполняется только одно из двух условий. В данном тесте сумма заказа (4500) меньше порога (5000), но покупатель является членом клуба (True). Логическое and требует истинности обеих частей, поэтому дает False. Логическому or достаточно выполнения хотя бы одного условия, поэтому оно возвращает True.

In [68]:
# Расширение 3. Промокод
has_promo_code = True
is_member = True

# Сформулируйте правило: можно ли объединять промокод и клубную скидку?
# Присвойте discount_rate и проверьте минимум три сочетания True / False.
# Сформулируем правило: скидки СКЛАДЫВАЮТСЯ, но максимальная суммарная скидка не может превышать 15%.
# Пусть промокод дает 10%, а клубная карта (is_member) дает 7%.
def calculate_discount(has_promo_code, is_member):
    discount_rate = 0.0
    if has_promo_code:
        discount_rate += 0.10
    if is_member:
        discount_rate += 0.07

    # Ограничение совместного применения
    if discount_rate > 0.15:
        discount_rate = 0.15
    return discount_rate

# Проверка 3 сочетаний True / False
print(f"Промокод: True, Карта: True -> Скидка: {calculate_discount(True, True)*100:.0f}%")
print(f"Промокод: True, Карта: False -> Скидка: {calculate_discount(True, False)*100:.0f}%")
print(f"Промокод: False, Карта: True -> Скидка: {calculate_discount(False, True)*100:.0f}%")

Промокод: True, Карта: True -> Скидка: 15%
Промокод: True, Карта: False -> Скидка: 10%
Промокод: False, Карта: True -> Скидка: 7%


**Вывод расширения 3:** Какое правило вы реализовали и к чему оно приводит? Какое правило вы реализовали и к чему оно приводит? Я реализовал правило кумулятивного сложения скидок с ограничением верхнего порога (лимита). Оно приводит к тому, что при раздельном использовании покупатель получает честные 10% или 7%, а при одновременном применении срабатывает лимит, защищающий магазин от чрезмерных убытков, ограничивая общую скидку значением 15% вместо 17%.

In [69]:
# Расширение 4. Собственное правило магазина
# Сначала запишите правило комментарием.
# Затем реализуйте его и проверьте минимум три набора данных.
# Правило: Если количество товаров (quantity) >= 5, то предоставляется
# оптовая скидка 15%. Иначе скидка составляет 0%.
def calculate_wholesale_discount(quantity):
    if quantity >= 5:
        discount_rate = 0.15
    else:
        discount_rate = 0.0
    return discount_rate
# Набор 1: Меньше порогового значения (4 штуки)
qty1 = 4
print(f"Количество: {qty1} шт. -> Скидка: {calculate_wholesale_discount(qty1)*100}%")

# Набор 2: Ровно на границе порога (5 штук)
qty2 = 5
print(f"Количество: {qty2} шт. -> Скидка: {calculate_wholesale_discount(qty2)*100}%")

# Набор 3: Больше порогового значения (10 штук)
qty3 = 10
print(f"Количество: {qty3} шт. -> Скидка: {calculate_wholesale_discount(qty3)*100}%")

Количество: 4 шт. -> Скидка: 0.0%
Количество: 5 шт. -> Скидка: 15.0%
Количество: 10 шт. -> Скидка: 15.0%


**Вывод расширения 4:** Почему выбраны именно эти проверки? Что они доказали? Выбранные проверки представляют собой классическое тестирование граничных значений: значение чуть меньше границы (4), точно на границе (5) и заведомо больше нее (10). Они доказали, что логическое условие quantity >= 5 работает корректно — скидка в 15% активируется строго начиная с 5 товаров, а при покупке 4 товаров дополнительная скидка не применяется.

## Проверка перед сдачей

- [+] Все обязательные задания выполнены.
- [+] Код выполняется сверху вниз без ошибок.
- [+] Заполнены вопрос и гипотеза.
- [+] Завершена базовая модель заказа.
- [+] Выполнены все четыре обязательных эксперимента.
- [+] Проверены значения ниже, на границе и выше границы.
- [+] В условиях изменяются значения скидки и доставки.
- [+] Выполнены минимум два задания расширенного уровня.
- [+] Для каждого эксперимента записан отдельный вывод.
- [+] Сформулирован общий вывод не менее чем из трёх предложений.
- [+] Итоговый результат выведен через f-строку.
- [+] Файл загружен как `homework/lesson-02/solution.ipynb`.